### ISSUE TITLE
Unsafe vector overflow

### ISSUE THREAD

**Issue by Veedrac** [rust-lang/rust#26550]
This gives undefined behaviour due to overflow:

    fn main() {
        let one_iter = Some(0).into_iter();
        let big_iter = std::iter::repeat(0).take(usize::max_value());
        
        one_iter.chain(big_iter).collect::<Vec<_>>();
    }

The problem is the overflow on the addition here:

https://github.com/rust-lang/rust/blob/a6430925b435d5c27bb7fb4088dee06817c83d87/src/libcollections/vec.rs#L1485-L1489

**Comment by alexcrichton** [comment]
Could you clarify what the undefined behavior you're seeing is? This is just a standard panic for me:

```
thread '<main>' panicked at 'capacity overflow', /home/rustbuild/src/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/option.rs:330
```

**Comment by Veedrac** [comment]
Forgot to mention, this only happens on nightly.

http://is.gd/OdmHAG

**Comment by sfackler** [comment]
The SIGILL is the result of the failure to malloc the entire address space.

I believe the only undefined behavior badness the `from_iter` logic can run into is if `1 + lower` overflows to 0, in which case the `get_unchecked_mut` call will end up as an out of bounds write.

**Comment by sfackler** [comment]
cc @Gankro

**Comment by Veedrac** [comment]
> I believe the only undefined behavior badness the `from_iter` logic can run into is if `1 + lower` overflows to 0, in which case the `get_unchecked_mut` call will end up as an out of bounds write.

That's what's happening.

I was expecting `oom` to give a different error, though. To evidence my claim, `gdb` says it crashes on a null pointer write:

    r::ptr::write<i32> (dst=0x0, src=-1) at ../src/libcore/ptr.rs:247
    247	../src/libcore/ptr.rs: No such file or directory.

whereas decreasing the memory a bit to get an OOM errors inside of `with_capacity`.

**Comment by Veedrac** [comment]
There's another unchecked addition here:

https://github.com/rust-lang/rust/blob/af9693cd9ee222b733d19499de97e767d649cf74/src/libcollections/vec.rs#L1573

**Comment by sfackler** [comment]
They're both fixed in #26563

**Comment by Veedrac** [comment]
Indeed, sorry.
